upb/def.c

```diff
@@ -1493,7 +1529,8 @@ bool upb_msgdef_addoneof(upb_msgdef *m, upb_oneofdef *o, const void *ref_donor,
 
 void upb_msgdef_setprimitiveshavepresence(upb_msgdef *m, bool have_presence) {
   assert(!upb_msgdef_isfrozen(m));
-  m->primitives_have_presence = have_presence;
+  assert(!m->base.file);
+  m->syntax = have_presence ? UPB_SYNTAX_PROTO2 : UPB_SYNTAX_PROTO3;
```

Comment:
I decided that having the msgdef field be called "syntax" makes the most sense, since this is how type.proto represents it (https://github.com/google/protobuf/blob/master/src/google/protobuf/type.proto#L58) and because a message's "syntax" affects more semantics than just field presence (for example, PROTO3 discards unknown fields by definition).

I think you're right though that this function can be removed. Done.